File: 31-next-permutation/next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        
        // 1. Find the pivot (first decreasing element from the right)
        int i = n - 2;
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }
        
        // If a pivot is found (i >= 0)
        if (i >= 0) {
            // 2. Find the swap element (smallest element greater than nums[i] in the suffix)
            int j = n - 1;
            while (j >= 0 && nums[j] <= nums[i]) {
                j--;
            }
            // Swap nums[i] and nums[j]
            swap(nums, i, j);
        }
        
        // 3. Reverse the suffix from i + 1 to the end
        reverse(nums, i + 1, n - 1);
    }
    
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }
}

File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int i = n - 2;

        // Step 1: Find first decreasing element from the right
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        // Step 2: If such element found, find just greater element to the right
        if (i >= 0) {
            int j = n - 1;
            while (nums[j] <= nums[i]) {
                j--;
            }
            // Swap nums[i] and nums[j]
            swap(nums[i], nums[j]);
        }

        // Step 3: Reverse the part from i+1 to end
        reverse(nums.begin() + i + 1, nums.end());
    }
};
